File: Final/Code/Project_4/BFS/General_Graph/General_Graph.py
from collections import deque, defaultdict

def bfs(graph, start):
    visited = set()
    queue = deque([start])
    visited.add(start)
    order = []

    while queue:
        node = queue.popleft()
        order.append(node)
        for neighbor in graph[node]:
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append(neighbor)
    return order

graph = defaultdict(list)
graph['A'].extend(['B', 'C'])
graph['B'].extend(['D', 'E'])
graph['C'].extend(['F'])
graph['D'].extend([])
graph['E'].extend(['F'])
graph['F'].extend([])

print("Thứ tự duyệt BFS:", bfs(graph, 'A'))

File: Final/Code/Project_4/BFS/General_Graph/General_Graph.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <queue>
#include <unordered_set>

using namespace std;

void bfs(unordered_map<char, vector<char>>& graph, char start) {
    unordered_set<char> visited;
    queue<char> q;
    q.push(start);
    visited.insert(start);

    while (!q.empty()) {
        char node = q.front();
        q.pop();
        cout << node << " ";

        for (char neighbor : graph[node]) {
            if (visited.find(neighbor) == visited.end()) {
                visited.insert(neighbor);
                q.push(neighbor);
            }
        }
    }
}

int main() {
    unordered_map<char, vector<char>> graph;
    graph['A'] = {'B', 'C'};
    graph['B'] = {'D', 'E'};
    graph['C'] = {'F'};
    graph['D'] = {};
    graph['E'] = {'F'};
    graph['F'] = {};

    cout << "Thứ tự duyệt BFS: ";
    bfs(graph, 'A');
    cout << endl;

    return 0;
}